<launch>
    <arg name="namespace" default="$(env UAV_NAMESPACE)"/>
    <arg name="odom"        default="mavros/global_position/local"/>

    <arg name="x_offset"    default="visual_servo/offset_x"/>
    <arg name="y_offset"    default="visual_servo/offset_y"/>
    <arg name="z_offset"    default="visual_servo/offset_z"/>
    <arg name="x_target_error" default="visual_servo/target_error_x"/>
    <arg name="y_target_error" default="visual_servo/target_error_y"/>
    <arg name="z_target_error" default="visual_servo/target_error_z"/>
    <arg name="process_value" default="visual_servo/process_value"/>
    <arg name="status"      default="visual_servo/is_enabled"/>

    <group ns="$(arg namespace)">
        
        <!-- Visual servo state machine parameters -->
        <rosparam command="load"
            file="$(find uav_ros_control)/config/reference/visual_servo_pursuit_sm_ch1.yaml" />

        <!-- Visual Servo state machine node -->
        <node
            pkg="uav_ros_control"
            type="wait_and_observe_sm"
            name="visual_servo_state_machine"
            output="screen">

            <remap from="odometry" to="$(arg odom)"/>
        </node>

                 <!-- Visual servo state machine parameters -->
        <rosparam command="load"
            file="$(find uav_ros_control)/config/reference/visual_servo_pursuit.yaml" />
        
        <!-- Visual servo node -->
        <node
            pkg="uav_ros_control"
            type="visual_servo_node"
            name="visual_servo_node"
            output="screen">

            <!-- Topic remapping -->
            <remap from="x_error" to="sm_pursuit/x_err"/>
            <remap from="y_error" to="sm_pursuit/y_err"/>
            <remap from="z_error" to="sm_pursuit/z_err"/>
            <remap from="yaw_error" to="sm_pursuit/yaw_err"/>
            <remap from="odometry" to="$(arg odom)"/>
            <remap from="imu" to="mavros/imu/data"/>

            <!-- State machine topics -->
            <remap from="x_offset" to="$(arg x_offset)"/>
            <remap from="y_offset" to="$(arg y_offset)"/>
            <remap from="z_offset" to="$(arg z_offset)"/>
            <remap from="visual_servo/x_error" to="$(arg x_target_error)"/>
            <remap from="visual_servo/y_error" to="$(arg y_target_error)"/>
            <remap from="visual_servo/z_error" to="$(arg z_target_error)"/>
            <remap from="VisualServoProcessValueTopic" to="$(arg process_value)"/>


        </node>
        
    </group>
</launch>
